packages/@aws-cdk/core/package.json

```diff
@@ -134,13 +134,14 @@
     "awslint": "cdk-awslint",
     "build+test+package": "npm run build+test && npm run package",
     "build+test": "npm run build && npm test",
-    "compat": "cdk-compat"
+    "compat": "cdk-compat",
+    "gen": "cfn2ts"
   },
   "cdk-build": {
     "cloudformation": "AWS::CloudFormation",
     "cfn2ts-core-import": ".",
     "pre": [
-      "rm -rf test/fs/fixtures && cd test/fs && tar -xzf fixtures.tar.gz"
+      "rm -rf test/fs/fixtures && cd test/fs && tar -xzf fixtures.tar.gz && npm run gen"
```

Comment:
@rix0rrr since `pre` is an array I assumed I can simply add `npm run gen` as a separate item in the array:

```
"pre": [
    "rm -rf test/fs/fixtures && cd test/fs && tar -xzf fixtures.tar.gz",
     "npm run gen"
] 
```
This does not work in[ `cdk-build`](https://github.com/aws/aws-cdk/blob/master/tools/cdk-build-tools/bin/cdk-build.ts#L33) as the array of commands is concatenated without an `&&`,  which results in:
```
rm -rf test/fs/fixtures && cd test/fs && tar -xzf fixtures.tar.gz  npm run gen
```
Which fails. Is this a bug?

Reply:
Oh yes, that sounds like a big bad bug.

Reply:
https://github.com/aws/aws-cdk/blob/d412bb6b67b4097b845e6dd9e0421e26ee2de458/tools/cdk-build-tools/lib/os.ts#L22-L24

I see that the implementation of `shell` is a little weird, probably because Node.js is just slightly different from other programming languages that have similar interfaces. Normally it's:

* Run a command through the shell, argument should be a string (string can contain special shell characters like `&`, `>` and they will have their usual shell meaning)
* Run a command directly, argument should be an array (shell is not involved, `&`, `>` and ` ` will just be interpreted as characters in a file name).

However, in inimitable "I'll try to do the best I can" Node.js fashion, the underlying `child_process.spawn` function takes an array yet runs it through the shell, and it's fine if the whole string is in `command[0]`. I guess in an attempt to be broadly compatible, it just won't escape `command[0]` but it will escape `command[1..]`. Something we need to verify.

> I consider that flavor of API to be a feature! It's much easier to get wrong by accidentally having shell metacharacters
> interpreted, or having to quote arguments (maybe there are filenames with spaces), etc. If I can, I will *always* design
> my subprocess API to take an array instead of a string. You can still get the original behavior by calling
> `subprocess(['bash', '-c', originalCommandStringWithSpecialCharacters])`, so it's not a loss of functionality.
>  It's a pity it's so confusing here.

In any case, knowing what we're knowing, and even leaving the API of the `shell` function unchanged, we can change the API exposed via the `package.json` file easily. If we detect that `package['cdk-build'].pre` is an array, we can either run multiple shells, or just join everything together with a ` && ` before we run it through the shell.

https://github.com/aws/aws-cdk/blob/d412bb6b67b4097b845e6dd9e0421e26ee2de458/tools/cdk-build-tools/bin/cdk-build.ts#L33

```ts
if (Array.isArray(options.pre)) {
  for (const cmd of options.pre) {
    await shell(cmd, { timers, env });
  }
} else {
  await shell(options.pre, { timers, env });
}
```